api/v1alpha1/azurevalidator_types.go

```diff
@@ -21,20 +21,42 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
 // AzureValidatorSpec defines the desired state of AzureValidator
 type AzureValidatorSpec struct {
-	Auth          AzureAuth `json:"auth"`
-	DefaultRegion string    `json:"defaultRegion"`
+	// Rules for validating role assignments in Azure RBAC.
+	RoleAssignmentRules []RoleAssignmentRule `json:"roleAssignmentRules"`
+}
+
+func (s AzureValidatorSpec) ResultCount() int {
+	return len(s.RoleAssignmentRules)
+}
 
-	// TODO: Add Azure specific stuff to be validated (probably just the role
-	// at first).
+// RoleAssignmentRule is a rule that validates that one or more desired role assignments exist
+// within a subscription. For each role assignment, the role is specified as its role name (e.g.
+// "Contributor") or its name (e.g. "b24988ac-6180-42a0-ab88-20f7382dd24c" for Contributor). If the
+// role name is specified, the validator takes care of looking up the name automatically.
+type RoleAssignmentRule struct {
+	Roles              []Role `json:"roles"`
+	ServicePrincipalID string `json:"servicePrincipalId"`
+	SubscriptionID     string `json:"subscriptionId"`
 }
 
-type AzureAuth struct {
-	// Option 1: lookup Azure creds from a secret
-	SecretName string `json:"secretName,omitempty"`
-	// Option 2: specify a service account (AKS)
-	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+// Role allow users to specify either a role's role name (e.g. "Contributor") or a role's name (e.g.
+// "b24988ac-6180-42a0-ab88-20f7382dd24c"), which is the name of the role with the role name
+// "Contributor". This allows role assignments with custom roles to be validated too, not just
+// built-in roles.
+//
+// If role is specified, it is used. If role is not specified but role name is specified, role name
+// is used. If neither are specified, it is a misconfiguration and validation will fail.
+//
+// TODO: Look into if we can use a webhook to validate this at apply time instead of allowing it to
```

Comment:
Please remove this TODO or move it somewhere else. It's landing in the description field of the CRD.

Reply:
Also, we should just let the validator fail. Adding a webhook to do additional validation would be overkill IMO. The validator is meant to fail if configuration is invalid. No need to add a second layer of validation in front of the validator.